graph/graph.go

```diff
@@ -0,0 +1,251 @@
+package graph
+
+import (
+	"fmt"
+	"strings"
+)
+
+// Graph is a map-basd adjacency list for storing a lineage graph
+type Graph struct {
+	graph map[IsID][]IsID
+	nodes map[IsID]Node
+}
+
+// IDfactory is an interface which creates new node IDs
+type IDfactory interface {
+	New() IsID
+}
+
+// IsID represents any type which can function as a node ID
+type IsID interface {
+	isID()
+}
+
+// ID is an auto-incrementing ID
+type ID struct {
+	currentID int
+}
+
+// New creates a new ID
+func (id *ID) New() IsID {
+	newID := id.currentID
+	id.currentID++
+	return intID(newID)
+}
+
+// intID is an auto-incrementing ID
```

Comment:
What's the diff between this and ID?

Reply:
ID is actually an id factory; ID is a method for creating new intIDs. I've changed its name to autoIncFactory, since it creates auto incrementing IDs.
intID is just a type of ID which uses ints as a basis.